File: pkm/widgets/filesystems.py
from pkm.widgets.base import BaseWidget
from pkm import utils

NEWLINE = '${voffset 10}\n'


class FileSystemsWidget(BaseWidget):
    """ Displays the NVIDIA gpu metrics. """

    def __init__(self, wsettings, origin=0):
        super().__init__(wsettings, origin)
        # Header and footer are 67px
        # Each filesystem row adds 36px
        # Subtract 10px no voffset on last device
        self.height = 57 + (len(self.filesystems) * 36)

    def get_conkyrc(self, theme):
        """ Create the conkyrc template for the this widget. """
        rows = []
        for fs in self.filesystems:
            rows.append(utils.clean_spaces(f"""
                ${{goto 10}}{theme.value}{fs['name']}${{alignr 55}}${{fs_free {fs['path']}}} free
                ${{goto 10}}{theme.value}${{fs_used_perc {fs['path']}}}%{theme.value}${{alignr 55}}${{fs_size {fs['path']}}} total
            """))
        graphfs = self.graph_fs or self.filesystems[0]['path']
        return utils.clean_spaces(f"""
            ${{voffset 17}}${{goto 100}}{theme.accent1}${{diskiograph {graphfs} 24,90}}
            ${{voffset -36}}${{goto 10}}{theme.header}File Systems
            ${{goto 10}}{theme.subheader}IO: ${{diskio {graphfs}}}/s
            ${{voffset 17}}{NEWLINE.join(rows)}
        """)  # noqa

    def get_lua_entries(self, theme):
        """ Create the draw.lua entries for this widget. """
        origin, width, height = self.origin, self.width, self.height
        return [
            self.draw('line', frm=(100,origin), to=(100,origin+40), thickness=width, color=theme.header_bg),  # header bg
            self.draw('line', frm=(100,origin+40), to=(100,origin+height), thickness=width, color=theme.bg),  # main bg
            self.draw('line', frm=(100,origin+20), to=(190, origin+20), thickness=24, color=theme.graph_bg),  # fs graph
        ] + list(self._fs_entries(theme, origin))

    def _fs_entries(self, theme, origin):
        y = origin+65
        for fs in self.filesystems:
            yield self.draw('ring_graph', conky_value=f"fs_used_perc {fs['path']}", center=(172,y),
                radius=8, bar_color=theme.accent1, bar_thickness=4, background_color=theme.graph_bg)
            y += 36

File: pkm/widgets/filesystems.lua
local config = require 'config'
local draw = require 'pkm/draw'
local socket = require 'socket'
local utils = require 'pkm/utils'

local filesystems = {}
filesystems.origin = 0
filesystems.height = 0
filesystems.history = nil

-- Draw
-- Draw this widget
function filesystems:draw()
  self.height = 57

  -- Header
  draw.widget_header(self.origin, self.height, 'File Systems', 'IO: '..self.diskio..'/s', function()
    draw.graph{data=self.history, x=100, y=self.origin+8, width=90, height=24, color=config.accent,
      bgcolor=config.header_graph_bg, minmaxvalue=100*1024, logscale=self.logscale}
  end)

  -- filesystems
  y = self.origin + 61
  for _, fs in ipairs(self.paths) do
    draw.rectangle{x=0, y=y-4, width=conky_window.width, height=40, color=config.background} -- background
    draw.text{x=10, y=y, text=fs.name, color=config.value} -- name
    draw.text{x=145, y=y, text=fs.free..' free', color=config.value, align='right'} -- fs free
    draw.text{x=10, y=y+15, text=fs.usedpct..'%', color=config.value} -- fs percent
    draw.text{x=145, y=y+15, text=fs.size..' total', color=config.value, align='right'} -- fs size
    draw.ringgraph{value=fs.usedpct, x=172, y=y+3, radius=9, width=5, color=config.accent, bgcolor=config.graph_bg} -- fs percent
    self.height = self.height + 40
    y = y + 40
  end
end

-- Update
-- Update IO History
function filesystems:update()
  if utils.check_update(self.last_update, self.update_interval) then
    -- Filesystem Info
    for _, fs in ipairs(self.paths) do
      fs.free = utils.parse('fs_free '..fs.path)
      fs.usedpct = tonumber(utils.parse('fs_used_perc '..fs.path))
      fs.size = utils.parse('fs_size '..fs.path)
    end
    -- Disk IO History
    self.diskio = utils.parse('diskio')
    local io, unit = self.diskio:match('^(%d+).*(%a+)$')
    io = tonumber(io)
    if unit == 'K' then io = io * 1024 end
    if unit == 'M' then io = io * 1024^2 end
    if unit == 'G' then io = io * 1024^3 end
    self.history = utils.push_history(self.history, io, 90)
    self.last_update = socket.gettime()
  end
end

-- Click
-- Perform click action
function filesystems:click(event, x, y)
  os.execute(self.onclick..' &')
end

return filesystems
